src/controls/adaptiveCardHost/AdaptiveCardHost.tsx

```diff
@@ -0,0 +1,232 @@
+    }, [props.card, props.onSetCustomElements, props.onSetCustomActions, props.onUpdateHostCapabilities]);
+    // *****
+
+    // create executeAction
+    const invokeAction = useCallback((action: Action) => {
+        if (props.onInvokeAction) {
+            let actionResult: IAdaptiveCardHostActionResult;
+            const type = action.getJsonTypeName();
+            switch (type) {
+                case OpenUrlAction.JsonTypeName: {
+                    let typedAction = action as OpenUrlAction;
+                    actionResult = {
+                        type: typedAction.getJsonTypeName(),
+                        title: typedAction.title,
+                        url: typedAction.url
+                    };
+                }
+                    break;
+
+                case SubmitAction.JsonTypeName: {
+                    let typedAction = action as SubmitAction;
+                    actionResult = {
+                        type: typedAction.getJsonTypeName(),
+                        title: typedAction.title,
+                        data: typedAction.data
+                    };
+                }
+                    break;
+                case ExecuteAction.JsonTypeName: {
+                    let typedAction = action as ExecuteAction;
+                    actionResult = {
+                        type: typedAction.getJsonTypeName(),
+                        title: typedAction.title,
+                        data: typedAction.data,
+                        verb: typedAction.verb
+                    };
+                }
+                    break;
+            }
+
+            props.onInvokeAction(actionResult);
+        }
+    }, [props.onInvokeAction]);
+    // *****
+
+    // set hostConfig
+    useEffect(() => {
+        let theme: PartialTheme | Theme;
+        let themeType = props.themeType;
+
+        if (!themeType) {
+            themeType = AdaptiveCardHostThemeType.SharePoint;
+        }
+
+        switch (themeType) {
+            case AdaptiveCardHostThemeType.SharePoint: {
+                let contextTheme = fluentUICustomizerContext.customizations.settings["theme"];
+                if (props.theme) {
+                    theme = props.theme;
+                } else if (contextTheme) {
+                    theme = contextTheme;
+                } else {
+                    theme = getDefaultFluentUITheme();
+                }
+            }
+                break;
+            case AdaptiveCardHostThemeType.Teams: {
+                theme = createDefaultTeamsTheme();
+            }
+                break;
+            case AdaptiveCardHostThemeType.TeamsDark: {
+                theme = createDarkTeamsTheme();
+            }
+                break;
+            case AdaptiveCardHostThemeType.TeamsHighContrast: {
+                theme = createHighContrastTeamsTheme();
+            }
+                break;
+        }
+
+        fluentUIThemeInstanceRef.current = convertFromPartialThemeToTheme(theme);
```

Comment:
Nope, the problems using vars is described here: https://www.emgoto.com/storing-values-with-useref/#whats-the-difference-between-useref-and-using-a-variable

Reply:
`ref`s are important to store the value between calls and to avoid re-call of `useEffect`.
But in your case here you do `fluentUIThemeInstanceRef.current = convertFromPartialThemeToTheme(theme)` and then 4 lines lower you are using `fluentUIThemeInstanceRef.current` which will not change in between.
So why not do that
```typescript
const themeInstance = converFromPartialThemetoTheme(theme);
fluentUIThemeInstanceRef.current = themeInstance;
// case block
hostConfig = <call_you_method>(themeInstance);
```

You can leave as is but variable should be more performant.

Reply:
@AJIXuMuK sorry I didn't understand your suggestion... ok, you are right, let me fix today 🙂

Reply:
Done 🙂